src/Unleash/Utilities/ToggleBootstrapUrlProvider.cs

```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Unleash.Internal;
+using Unleash.Logging;
+using Unleash.Serialization;
+
+namespace Unleash.Utilities
+{
+    public class ToggleBootstrapUrlProvider : IToggleBootstrapProvider
+    {
+        private static readonly ILog Logger = LogProvider.GetLogger(typeof(ToggleBootstrapUrlProvider));
+
+        private readonly CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+        private readonly HttpClient client;
+        private readonly IJsonSerializer jsonSerializer;
+        private readonly string path;
+        private readonly bool throwOnFail;
+        private readonly Dictionary<string, string> customHeaders;
+
+        public ToggleBootstrapUrlProvider(string path, HttpClient client, IJsonSerializer jsonSerializer, bool throwOnFail = false, Dictionary<string, string> customHeaders = null)
+        {
+            this.path = path;
+            this.client = client;
+            this.jsonSerializer = jsonSerializer;
+            this.throwOnFail = throwOnFail;
+            this.customHeaders = customHeaders;
+        }
+
+        public ToggleCollection Read()
+        {
+            return Task.Run(() => FetchFile()).GetAwaiter().GetResult();
+        }
+
+        private async Task<ToggleCollection> FetchFile()
+        {
+            using (var request = new HttpRequestMessage(HttpMethod.Get, path))
+            {
+                if (customHeaders != null)
+                {
+                    foreach (var keyValuePair in customHeaders)
+                    {
+                        request.Headers.Add(keyValuePair.Key, keyValuePair.Value);
+                    }
+                }
+
+                using (var response = await client.SendAsync(request, cancellationTokenSource.Token).ConfigureAwait(false))
+                {
+                    if (!response.IsSuccessStatusCode)
+                    {
+                        var error = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+                        Logger.Trace($"UNLEASH: Error {response.StatusCode} from server in 'ToggleBootstrapUrlProvider.{nameof(FetchFile)}': " + error);
+
+                        if (throwOnFail)
+                            throw new FetchingToggleBootstrapUrlFailedException("Failed to fetch feature toggles", response.StatusCode);
+
+                        return null;
+                    }
+
+                    var togglesResponseStream = await response.Content.ReadAsStreamAsync();
+                    return jsonSerializer.Deserialize<ToggleCollection>(togglesResponseStream);
```

Comment:
It might be worth wrapping these two lines in some kind of error handling. Since we have a neat little option to configure whether or not the SDK will raise an exception when bootstrapping fails, I'd be quite surprised to set this to false and then get an exception. ReadAsStreamAsync can fail in a few cases, the notable one for me is that when Unleash servers are extremely overloaded, they may choose to close the connection early, this will raise an exception here and crash(?) the parent process. The deserialize can also fail, that's probably more on the user for pointing the bootstrap to a broken data source so open to opinions on that one but I do think we should trap the first case

Reply:
@sighphyre Ok fixed. I figured wrap both, if we're only trapping the first one then we should name the parameter accordingly to avoid confusion, so opted to trap both cases. Personal opinion, but open to discussion.

Reply:
I like it!